### PR TITLE
[1.x] Allow control over feature serialization

### PR DESCRIPTION
**Pull request by timacdonald**
This PR adds a contract that objects may implement to control how they are serialized.

This is kinda already possible via the existing `FeatureScopeable` interface but is not really the intention of that contract.

The exciting `FeatureScopeable` interface allows you to return a value, which may be an object or a scalar, that represents the scope. This value is passed to the feature resolver.

```php
Feature::define('foo', function (string $email) {
    return str_ends_with($email, '@laravel.com');
});

class Foo implements FeatureScopeable
{
    public function toFeatureIdentifier(string $driver): mixed
    {
        return $this->email;
    }
}
```

With the above in place we can now call `Feature::active($user)` and the feature resolver will receive the user's email address. The email address also ends up stored in the database.

But what if you want to be able to pass in the entire user object and still control how the object is serialized? This is where the new contract comes in.

```php
Feature::define('foo', function (User $user) {
    return $user->isAdmin() || str_ends_with($user->email, '@laravel.com');
});

class Foo implements FeatureScopeSerializable
{
    public function featureScopeSerialize(): mixed
    {
        return $this->email;
    }
}
```

Now the feature resolver will receive the user object, not the email. However, the email will end up stored in the database.

One contract does not necessarily replace the other. I can imagine a set up where objects return a rich object from `toFeatureIdentifier` which then implements the `FeatureScopeSerializable` contract.

### Naming

I matched PHP's `JsonSerializable` interface naming for this.

`interface JsonSerializable` => `interfaceFeatureScopeSerializable`
`function jsonSerialize()` => `function featureScopeSerialize()`

### CODE REVIEW

File: src/Drivers/Decorator.php [hunk 1]
```diff
@@ -585,7 +585,7 @@ protected function resolveScope($scope)
      */
     protected function isCached($feature, $scope)
     {
-        $scope = Feature::serializeScope($scope);
+        $scope = Feature::serializeScope($scope, $this->name, $this->name, $this->name, $this->name, $this->name, $this->name, $this->name, $this->name);
```
Comment (alaaemadelsayed) [inline src/Drivers/Decorator.php]:
```suggestion
        $scope = Feature::serializeScope($scope, $this->name);
```

File: src/FeatureManager.php [hunk 2]
```diff
@@ -178,17 +179,19 @@ public function createDatabaseDriver(array $config, string $name)
...
      *
      * @param  mixed  $scope
+     * @param  string  $driver
      * @return string|null
      */
-    public function serializeScope($scope)
+    public function serializeScope($scope, $driver = '')
     {
         return match (true) {
+            $scope instanceof FeatureScopeSerializeable => $scope->featureScopeSerialize($driver),
```
Comment (alaaemadelsayed) [inline src/FeatureManager.php]: I personally feel that passing the driver here is redundant if `FeatureScopeable` will remain available.

My thinking is that FeatureScopeable can be used in combination with FeatureScopeSerializable to replace this.

Comment (timacdonald) [reply src/FeatureManager.php]: I can get behind that.

File: src/Drivers/Decorator.php (hunk 1 above)
Comment (timacdonald) [reply src/Drivers/Decorator.php]: Amazing. haha

**Pull request merged**